## Imports

In [3]:
import time
from timeit import default_timer as timer
import numpy as np
from functools import wraps
import cProfile
import subprocess
NAME = "Phoebe"

# getAcc Profiling Comparisons

In [6]:
# getAcc() 
! python -m memory_profiler sph.py

Filename: sph.py

Line #    Mem usage    Increment  Occurrences   Line Contents
   110  122.418 MiB -14139.141 MiB         301   @profile
   111                                         def getAcc( pos, vel, m, h, k, n, lmbda, nu ):
   112                                         	"""
   113                                         	Calculate the acceleration on each SPH particle
   114                                         	pos   is an N x 3 matrix of positions
   115                                         	vel   is an N x 3 matrix of velocities
   116                                         	m     is the particle mass
   117                                         	h     is the smoothing length
   118                                         	k     equation of state constant
   119                                         	n     polytropic index
   120                                         	lmbda external force constant
   121                                         	nu    viscosity


In [7]:
# Opt getAcc()
! python -m memory_profiler opt_sph.py

Filename: opt_sph.py

Line #    Mem usage    Increment  Occurrences   Line Contents
   109  122.039 MiB -14633.027 MiB         301   @profile
   110                                         def getAcc( pos, vel, m, h, k, n, lmbda, nu ):
   111                                         	"""
   112                                         	Calculate the acceleration on each SPH particle
   113                                         	pos   is an N x 3 matrix of positions
   114                                         	vel   is an N x 3 matrix of velocities
   115                                         	m     is the particle mass
   116                                         	h     is the smoothing length
   117                                         	k     equation of state constant
   118                                         	n     polytropic index
   119                                         	lmbda external force constant
   120                                         	nu    viscos

# getPressure Profiling Comparisons

In [ ]:
# get pressure
! python -m memory_profiler sph.py

# getDensity Profiling Comparisons

In [ ]:
# get density
! python -m memory_profiler sph.py

# pairwise Profiling Comparisons

In [18]:
# pairwise 
! python -m memory_profiler sph.py

Filename: sph.py

Line #    Mem usage    Increment  Occurrences   Line Contents
    48  118.188 MiB -7117.578 MiB        1202   @profile	
    49                                         def getPairwiseSeparations( ri, rj ):
    50                                         	"""
    51                                         	Get pairwise desprations between 2 sets of coordinates
    52                                         	ri    is an M x 3 matrix of positions
    53                                         	rj    is an N x 3 matrix of positions
    54                                         	dx, dy, dz   are M x N matrices of separations
    55                                         	"""
    56                                         	
    57  118.188 MiB -2831.113 MiB        1202   	M = ri.shape[0]
    58  118.188 MiB -2831.113 MiB        1202   	N = rj.shape[0]
    59                                         	
    60                                         	# positions ri = (x,y,z)
  

In [22]:
# pairwise Optimized
! python -m memory_profiler opt_sph.py

Filename: opt_sph.py

Line #    Mem usage    Increment  Occurrences   Line Contents
    60  118.648 MiB -7828.898 MiB        1202   @profile	
    61                                         def getPairwiseSeparations( ri, rj ):
    62                                         	# OPTIMIZED VERSION
    63                                         	"""
    64                                         	Get pairwise desprations between 2 sets of coordinates
    65                                         	ri    is an M x 3 matrix of positions
    66                                         	rj    is an N x 3 matrix of positions
    67                                         	dx, dy, dz   are M x N matrices of separations
    68                                         	"""
    69                                         
    70  119.871 MiB -2437.258 MiB        1202   	dx = np.subtract.outer(ri[:, 0], rj[:, 0])
    71  121.094 MiB -2801.922 MiB        1202   	dy = np.subtract.outer(ri[:, 1], rj[:, 1])

## getPairwise Optimization Results

Version  |Initial (MiB)|Max (MiB)|Final (MiB)   
---------|-------------|---------|----------
Original |   118.590   | 121.098 | 121.098 
Optimized|   119.555   | 216.289 | 125.668



# gradW Profiling Comparisons

In [9]:
# gradW Original
! python -m memory_profiler sph.py

Filename: sph.py

Line #    Mem usage    Increment  Occurrences   Line Contents
    28  119.438 MiB -2070.977 MiB         301   @profile	
    29                                         def gradW( x, y, z, h ):
    30                                         	"""
    31                                         	Gradient of the Gausssian Smoothing kernel (3D)
    32                                         	x     is a vector/matrix of x positions
    33                                         	y     is a vector/matrix of y positions
    34                                         	z     is a vector/matrix of z positions
    35                                         	h     is the smoothing length
    36                                         	wx, wy, wz     is the evaluated gradient
    37                                         	"""
    38                                         	
    39  123.105 MiB  747.633 MiB         301   	r = np.sqrt(x**2 + y**2 + z**2)
    40                        

In [13]:
# gradW Optimized
! python -m memory_profiler opt_sph.py

Filename: opt_sph.py

Line #    Mem usage    Increment  Occurrences   Line Contents
    28  119.555 MiB -1988.168 MiB         301   @profile	
    29                                         def gradW( x, y, z, h ):
    30                                         	"""
    31                                         	Gradient of the Gausssian Smoothing kernel (3D)
    32                                         	x     is a vector/matrix of x positions
    33                                         	y     is a vector/matrix of y positions
    34                                         	z     is a vector/matrix of z positions
    35                                         	h     is the smoothing length
    36                                         	wx, wy, wz     is the evaluated gradient
    37                                         	"""
    38                                         	# initalize the arrays
    39  119.555 MiB -273.734 MiB         301   	r = np.empty(x.shape, dtype=x.dtype)

## gradW Optimization Results


Version  |Initial (MiB)|Max (MiB)|Final (MiB)   
---------|-------------|---------|----------
Original |   119.438   | 870.738 | 125.551 
Optimized|   119.555   | 216.289 | 125.668

By optimizing the gradW function with np.empty to initailze arrays and using in-place operations we were able to significantly reduce the maximum memory usage during this function which increases the more efficient use of memory allocation. 

# W Profiling Comparisons

In [ ]:
# W
! python -m memory_profiler sph.py

# main Profiling Comparisons

In [ ]:
# main
! python -m memory_profiler sph.py

# Mprof 

In [12]:
! python -m mprof run opt_sph.py

mprof.py: Sampling memory every 0.1s
running new process
running as a Python program...


In [1]:
! python -m mprof plot mem_opt/mprofile_20250223123640.dat

Input file not found: mem_opt/mprofile_20250223123640.dat
No files found from given input.


# mprof notes
Original has a max usage of 10Mib in the first 0 seconds, and then stays steady. Then the 